go/genkit/session.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package genkit
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sync"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/firebase/genkit/go/internal/base"
+	"github.com/google/uuid"
+	"github.com/invopop/jsonschema"
+)
+
+// Sessions are by default stored in memory only.
+// Implement the SessionStore interface to persist state.
+type SessionStore interface {
+	Get(sessionId string) (data SessionData, err error)
+	Save(sessionId string, data SessionData) error
+}
+
+type SessionData struct {
+	// Any state that should be stored
+	State map[string]any `json:"state,omitempty"`
+	// Schema for state variables
+	StateSchema *jsonschema.Schema `json:"stateschema,omitempty"`
+	// The messages for each thread
+	Threads map[string][]*ai.Message `json:"threads,omitempty"`
+}
+
+type Session struct {
+	// Genkit
+	Genkit *Genkit
+	// The session id
+	ID string
+	// The data for the session
+	SessionData SessionData
+	// Default state variable values
+	DefaultState map[string]any
```

Comment:
Do we need this field? Can we just set `Data` with the default when we construct the `Session`?